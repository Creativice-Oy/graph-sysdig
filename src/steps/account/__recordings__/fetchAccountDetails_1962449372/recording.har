{
  "log": {
    "_recordingName": "fetchAccountDetails",
    "creator": {
      "comment": "persister:JupiterOneIntegationFSPersister",
      "name": "Polly.JS",
      "version": "5.1.1"
    },
    "entries": [
      {
        "_id": "0d29cdc1ea9a8e8f7ab793f1e65b94a7",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "application/json"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "app.us4.sysdig.com"
            }
          ],
          "headersSize": 285,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [],
          "url": "https://app.us4.sysdig.com/api/user/me"
        },
        "response": {
          "bodySize": 2675,
          "content": {
            "mimeType": "application/json;charset=utf-8",
            "size": 2675,
            "text": "{\"user\":{\"termsAndConditions\":true,\"timezone\":\"+00:00\",\"pictureUrl\":\"http://www.gravatar.com/avatar/ba5b87ffa4527a556b5f6da95fe36fe5\",\"customerSettings\":{\"sysdig\":{\"enabled\":false,\"enabledSSE\":false,\"buckets\":[]},\"plan\":{\"maxAgents\":15,\"onDemandAgents\":0,\"maxTeams\":-1,\"timelines\":[{\"from\":null,\"to\":null,\"sampling\":10000000},{\"from\":null,\"to\":null,\"sampling\":60000000},{\"from\":null,\"to\":null,\"sampling\":600000000},{\"from\":null,\"to\":null,\"sampling\":3600000000},{\"from\":null,\"to\":null,\"sampling\":86400000000}],\"metricsSettings\":{\"enforce\":false,\"showExperimentals\":false,\"limits\":{\"jmx\":500,\"statsd\":1000,\"appCheck\":500,\"prometheus\":8000,\"prometheusPerProcess\":500,\"connections\":80,\"progAggregationCount\":12,\"appCheckAggregationCount\":12,\"promMetricsWeight\":0.0,\"topFilesCount\":10,\"topDevicesCount\":10,\"hostServerPorts\":10,\"containerServerPorts\":5,\"limitKubernetesResources\":false,\"kubernetesPods\":10000,\"kubernetesJobs\":10000,\"containerDensity\":200,\"meerkatSuited\":true},\"enforceAgentAggregation\":true,\"enablePromCalculatedIngestion\":false,\"enablePromScrapeV2\":true,\"enableGlobalK8sAllPodContainers\":false},\"secureEnabled\":true,\"monitorEnabled\":false,\"allocatedAgentsCount\":15,\"subscriptionState\":\"trialing\",\"paymentsIntegrationId\":{\"id\":\"not present\",\"ttl\":{\"ttl\":43}},\"pricingPlan\":\"pro\",\"indirectCustomer\":false,\"trialPlanName\":\"secure-enterprise-standard-monthly\",\"partner\":\"Direct\",\"trialEndDate\":1664383443236,\"subscriptionVersion\":2,\"migratedToV2Direct\":false,\"overageAssessmentEligible\":false},\"environment\":{}},\"customer\":{\"id\":5001686,\"name\":null,\"accessKey\":\"1be2fe7d-f00f-4513-9a15-1384c35d8497\",\"externalId\":\"fbd8b3fc-781c-4e7f-afac-a2ee9309f4c9\",\"dateCreated\":1661791442836},\"oauth\":false,\"agentInstallParams\":{\"accessKey\":\"1be2fe7d-f00f-4513-9a15-1384c35d8497\",\"collectorAddress\":\"ingest.us4.sysdig.com\",\"collectorPort\":6443,\"checkCertificate\":true,\"sslEnabled\":true},\"remoteWriteParams\":{\"serverAddress\":\"app.us4.sysdig.com\"},\"properties\":{\"has_been_invited\":false},\"resetPassword\":false,\"additionalRoles\":[],\"teamRoles\":[{\"teamId\":50006626,\"teamName\":\"Secure Operations\",\"teamTheme\":\"#7BB0B2\",\"userId\":50003375,\"userName\":\"stefan.virag@contractor.jupiterone.com\",\"role\":\"ROLE_TEAM_MANAGER\",\"admin\":true},{\"teamId\":50006625,\"teamName\":\"Monitor Operations\",\"teamTheme\":\"#7BB0B2\",\"userId\":50003375,\"userName\":\"stefan.virag@contractor.jupiterone.com\",\"role\":\"ROLE_TEAM_MANAGER\",\"admin\":true}],\"lastUpdated\":1663493736587,\"dateActivated\":1661791669709,\"accessKey\":\"1be2fe7d-f00f-4513-9a15-1384c35d8497\",\"intercomUserIdHash\":\"7941192844d448aedd650df4bd26799efd5fca275a1f1eef775ee67b6cf061ad\",\"uniqueIntercomUserId\":\"50003375.fbd8b3fc-781c-4e7f-afac-a2ee9309f4c9\",\"currentTeam\":50006626,\"name\":\"stefan.virag@contractor.jupiterone.com\",\"id\":50003375,\"version\":2,\"enabled\":true,\"firstName\":\"Stefan\",\"lastName\":\"Virag\",\"systemRole\":\"ROLE_CUSTOMER\",\"username\":\"stefan.virag@contractor.jupiterone.com\",\"status\":\"confirmed\",\"products\":[\"SDC\",\"SDS\"],\"dateCreated\":1661791442936,\"lastSeenOnSecure\":1663493736587}}"
          },
          "cookies": [
            {
              "httpOnly": true,
              "name": "INGRESSCOOKIEAPI",
              "path": "/",
              "value": "[REDACTED]"
            }
          ],
          "headers": [
            {
              "name": "date",
              "value": "Sun, 18 Sep 2022 09:46:17 GMT"
            },
            {
              "name": "vary",
              "value": "Origin, Access-Control-Request-Method, Access-Control-Request-Headers, Accept-Encoding, User-Agent"
            },
            {
              "name": "content-type",
              "value": "application/json;charset=utf-8"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "cache-control",
              "value": "no-cache, no-store, max-age=0, must-revalidate"
            },
            {
              "name": "pragma",
              "value": "no-cache"
            },
            {
              "name": "expires",
              "value": "0"
            },
            {
              "name": "strict-transport-security",
              "value": "max-age=15768000"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "name": "connection",
              "value": "close"
            }
          ],
          "headersSize": 498,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2022-09-18T09:46:16.712Z",
        "time": 518,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 518
        }
      }
    ],
    "pages": [],
    "version": "1.2"
  }
}
