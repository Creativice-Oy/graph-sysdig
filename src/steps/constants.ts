import {
  RelationshipClass,
  StepEntityMetadata,
  StepRelationshipMetadata,
} from '@jupiterone/integration-sdk-core';

export const Steps = {
  ACCOUNT: 'fetch-account-details',
  USERS: 'fetch-users',
  BUILD_ACCOUNT_AND_USER_RELATIONSHIP: 'build-account-and-user-relationship',
  TEAMS: 'fetch-teams',
  IMAGE_SCANS: 'fetch-image-scans',
  BUILD_ACCOUNT_AND_IMAGE_SCAN_RELATIONSHIP:
    'build-account-and-image-scan-relationship',
  BUILD_ACCOUNT_AND_TEAM_RELATIONSHIP: 'build-account-and-team-relationship',
  BUILD_TEAM_AND_USER_RELATIONSHIP: 'build-team-and-user-relationship',
};

export const Entities: Record<
  'ACCOUNT' | 'USER' | 'TEAM' | 'IMAGE_SCAN',
  StepEntityMetadata
> = {
  ACCOUNT: {
    resourceName: 'Account',
    _type: 'sysdig_account',
    _class: ['Account'],
  },
  USER: {
    resourceName: 'User',
    _type: 'sysdig_user',
    _class: ['User'],
  },
  TEAM: {
    resourceName: 'Team',
    _type: 'sysdig_team',
    _class: ['Team'],
  },
  IMAGE_SCAN: {
    resourceName: 'Image Scan',
    _type: 'sysdig_image_scan',
    _class: ['Assessment'],
  },
};

export const Relationships: Record<
  | 'ACCOUNT_HAS_USER'
  | 'ACCOUNT_HAS_TEAM'
  | 'TEAM_HAS_USER'
  | 'ACCOUNT_HAS_IMAGE_SCAN',
  StepRelationshipMetadata
> = {
  ACCOUNT_HAS_USER: {
    _type: 'sysdig_account_has_user',
    sourceType: Entities.ACCOUNT._type,
    _class: RelationshipClass.HAS,
    targetType: Entities.USER._type,
  },
  ACCOUNT_HAS_TEAM: {
    _type: 'sysdig_account_has_team',
    sourceType: Entities.ACCOUNT._type,
    _class: RelationshipClass.HAS,
    targetType: Entities.TEAM._type,
  },
  TEAM_HAS_USER: {
    _type: 'sysdig_team_has_user',
    sourceType: Entities.TEAM._type,
    _class: RelationshipClass.HAS,
    targetType: Entities.USER._type,
  },
  ACCOUNT_HAS_IMAGE_SCAN: {
    _type: 'sysdig_account_has_image_scan',
    sourceType: Entities.ACCOUNT._type,
    _class: RelationshipClass.HAS,
    targetType: Entities.IMAGE_SCAN._type,
  },
};
