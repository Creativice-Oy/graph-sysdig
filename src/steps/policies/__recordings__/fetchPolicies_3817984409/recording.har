{
  "log": {
    "_recordingName": "fetchPolicies",
    "creator": {
      "comment": "persister:JupiterOneIntegationFSPersister",
      "name": "Polly.JS",
      "version": "5.1.1"
    },
    "entries": [
      {
        "_id": "0d29cdc1ea9a8e8f7ab793f1e65b94a7",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "application/json"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "app.us4.sysdig.com"
            }
          ],
          "headersSize": 285,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [],
          "url": "https://app.us4.sysdig.com/api/user/me"
        },
        "response": {
          "bodySize": 2673,
          "content": {
            "mimeType": "application/json;charset=utf-8",
            "size": 2673,
            "text": "{\"user\":{\"termsAndConditions\":true,\"timezone\":\"+00:00\",\"pictureUrl\":\"http://www.gravatar.com/avatar/ba5b87ffa4527a556b5f6da95fe36fe5\",\"customerSettings\":{\"sysdig\":{\"enabled\":false,\"enabledSSE\":false,\"buckets\":[]},\"plan\":{\"maxAgents\":15,\"onDemandAgents\":0,\"maxTeams\":-1,\"timelines\":[{\"from\":null,\"to\":null,\"sampling\":10000000},{\"from\":null,\"to\":null,\"sampling\":60000000},{\"from\":null,\"to\":null,\"sampling\":600000000},{\"from\":null,\"to\":null,\"sampling\":3600000000},{\"from\":null,\"to\":null,\"sampling\":86400000000}],\"metricsSettings\":{\"enforce\":false,\"showExperimentals\":false,\"limits\":{\"jmx\":500,\"statsd\":1000,\"appCheck\":500,\"prometheus\":8000,\"prometheusPerProcess\":500,\"connections\":80,\"progAggregationCount\":12,\"appCheckAggregationCount\":12,\"promMetricsWeight\":0.0,\"topFilesCount\":10,\"topDevicesCount\":10,\"hostServerPorts\":10,\"containerServerPorts\":5,\"limitKubernetesResources\":false,\"kubernetesPods\":10000,\"kubernetesJobs\":10000,\"containerDensity\":200,\"meerkatSuited\":true},\"enforceAgentAggregation\":true,\"enablePromCalculatedIngestion\":false,\"enablePromScrapeV2\":true,\"enableGlobalK8sAllPodContainers\":false},\"secureEnabled\":true,\"monitorEnabled\":false,\"allocatedAgentsCount\":15,\"subscriptionState\":\"trialing\",\"paymentsIntegrationId\":{\"id\":\"not present\",\"ttl\":{\"ttl\":43}},\"pricingPlan\":\"pro\",\"indirectCustomer\":false,\"trialPlanName\":\"secure-enterprise-standard-monthly\",\"partner\":\"Direct\",\"trialEndDate\":1664383443236,\"subscriptionVersion\":2,\"migratedToV2Direct\":false,\"overageAssessmentEligible\":false},\"environment\":{}},\"customer\":{\"id\":5001686,\"name\":null,\"accessKey\":\"1be2fe7d-f00f-4513-9a15-1384c35d8497\",\"externalId\":\"fbd8b3fc-781c-4e7f-afac-a2ee9309f4c9\",\"dateCreated\":1661791442836},\"oauth\":false,\"agentInstallParams\":{\"accessKey\":\"1be2fe7d-f00f-4513-9a15-1384c35d8497\",\"collectorAddress\":\"ingest.us4.sysdig.com\",\"collectorPort\":6443,\"checkCertificate\":true,\"sslEnabled\":true},\"remoteWriteParams\":{\"serverAddress\":\"app.us4.sysdig.com\"},\"properties\":{\"has_been_invited\":false},\"resetPassword\":false,\"additionalRoles\":[],\"teamRoles\":[{\"teamId\":50006626,\"teamName\":\"Secure Operations\",\"teamTheme\":\"#7BB0B2\",\"userId\":50003375,\"userName\":\"stefan.virag@contractor.jupiterone.com\",\"role\":\"ROLE_TEAM_MANAGER\",\"admin\":true},{\"teamId\":50006625,\"teamName\":\"Monitor Operations\",\"teamTheme\":\"#7BB0B2\",\"userId\":50003375,\"userName\":\"stefan.virag@contractor.jupiterone.com\",\"role\":\"ROLE_TEAM_MANAGER\",\"admin\":true}],\"lastUpdated\":1663575912770,\"dateActivated\":1661791669709,\"accessKey\":\"1be2fe7d-f00f-4513-9a15-1384c35d8497\",\"intercomUserIdHash\":\"7941192844d448aedd650df4bd26799efd5fca275a1f1eef775ee67b6cf061ad\",\"uniqueIntercomUserId\":\"50003375.fbd8b3fc-781c-4e7f-afac-a2ee9309f4c9\",\"currentTeam\":50006626,\"name\":\"stefan.virag@contractor.jupiterone.com\",\"id\":50003375,\"version\":2,\"enabled\":true,\"username\":\"stefan.virag@contractor.jupiterone.com\",\"products\":[\"SDS\",\"SDC\"],\"firstName\":\"Stefan\",\"lastName\":\"Virag\",\"systemRole\":\"ROLE_CUSTOMER\",\"status\":\"confirmed\",\"dateCreated\":1661791442936,\"lastSeenOnSecure\":1663575912770}}"
          },
          "cookies": [
            {
              "httpOnly": true,
              "name": "INGRESSCOOKIEAPI",
              "path": "/",
              "value": "[REDACTED]"
            }
          ],
          "headers": [
            {
              "name": "date",
              "value": "Mon, 19 Sep 2022 08:35:02 GMT"
            },
            {
              "name": "vary",
              "value": "Origin, Access-Control-Request-Method, Access-Control-Request-Headers, Accept-Encoding, User-Agent"
            },
            {
              "name": "content-type",
              "value": "application/json;charset=utf-8"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "cache-control",
              "value": "no-cache, no-store, max-age=0, must-revalidate"
            },
            {
              "name": "pragma",
              "value": "no-cache"
            },
            {
              "name": "expires",
              "value": "0"
            },
            {
              "name": "strict-transport-security",
              "value": "max-age=15768000"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "name": "connection",
              "value": "close"
            }
          ],
          "headersSize": 498,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2022-09-19T08:35:02.232Z",
        "time": 547,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 547
        }
      },
      {
        "_id": "6fd4e7a6bed8615501f32ff2a2f8486e",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "application/json"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "app.us4.sysdig.com"
            }
          ],
          "headersSize": 307,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [],
          "url": "https://app.us4.sysdig.com/api/scanning/policies/v2/policies"
        },
        "response": {
          "bodySize": 1684,
          "content": {
            "mimeType": "application/json",
            "size": 1684,
            "text": "{\"policies\":[{\"id\":17384,\"name\":\"Sysdig Best Practices\",\"identifier\":\"sysdig_best_practices\",\"description\":\"Recommended out of the box image scanning checks.\",\"policyType\":\"alwaysApply\",\"bundles\":[{\"id\":1,\"name\":\"Severe vulnerabilities with a Fix\",\"identifier\":\"severe_vulnerabilities_with_a_fix\",\"description\":\"Fail on Severe vulnerabilities that can be fixed.\",\"bundleType\":\"predefined\",\"rules\":[{\"ruleType\":\"vulnSeverityAndThreats\",\"predicates\":[{\"type\":\"vulnSeverity\",\"extra\":{\"level\":\"critical\"}},{\"type\":\"vulnIsFixable\"}]},{\"ruleType\":\"vulnSeverityAndThreats\",\"predicates\":[{\"type\":\"vulnSeverity\",\"extra\":{\"level\":\"high\"}},{\"type\":\"vulnIsFixableWithAge\",\"extra\":{\"age\":30}}]},{\"ruleType\":\"vulnSeverityAndThreats\",\"predicates\":[{\"type\":\"vulnSeverity\",\"extra\":{\"level\":\"high\"}},{\"type\":\"vulnExploitableViaNetwork\"}]}],\"creationTimestamp\":\"2022-03-08T16:42:57.270495Z\",\"updateTimestamp\":\"2022-03-08T16:42:57.270495Z\"}],\"creationTimestamp\":\"2022-09-02T11:58:59.03062Z\",\"updateTimestamp\":\"2022-09-02T11:58:59.03062Z\"},{\"id\":22461,\"name\":\"My New Pipeline Policy 1\",\"identifier\":\"my-new-pipeline-policy-1\",\"description\":\"My New Pipeline Policy Description 1\",\"policyType\":\"alwaysApply\",\"bundles\":[{\"id\":10803,\"name\":\"My new bundle 1\",\"identifier\":\"my-new-bundle-1\",\"description\":\"My new bundle description 1\",\"bundleType\":\"custom\",\"rules\":[{\"ruleType\":\"vulnSeverityAndThreats\",\"predicates\":[{\"type\":\"vulnSeverity\",\"extra\":{\"level\":\"critical\"}},{\"type\":\"vulnAge\",\"extra\":{\"age\":7}}]}],\"creationTimestamp\":\"2022-09-16T09:52:17.357177Z\",\"updateTimestamp\":\"2022-09-16T09:52:17.357177Z\"}],\"creationTimestamp\":\"2022-09-16T09:53:16.483187Z\",\"updateTimestamp\":\"2022-09-16T09:53:16.483187Z\"}]}\n"
          },
          "cookies": [
            {
              "httpOnly": true,
              "name": "INGRESSCOOKIEAPI",
              "path": "/",
              "value": "[REDACTED]"
            }
          ],
          "headers": [
            {
              "name": "content-type",
              "value": "application/json"
            },
            {
              "name": "x-ratelimit-limit",
              "value": "60"
            },
            {
              "name": "x-ratelimit-remaining",
              "value": "57"
            },
            {
              "name": "x-ratelimit-reset",
              "value": "1663576512"
            },
            {
              "name": "date",
              "value": "Mon, 19 Sep 2022 08:35:03 GMT"
            },
            {
              "name": "content-length",
              "value": "1684"
            },
            {
              "name": "strict-transport-security",
              "value": "max-age=15768000"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "name": "cache-control",
              "value": "private"
            },
            {
              "name": "connection",
              "value": "close"
            }
          ],
          "headersSize": 327,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2022-09-19T08:35:02.834Z",
        "time": 470,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 470
        }
      }
    ],
    "pages": [],
    "version": "1.2"
  }
}
