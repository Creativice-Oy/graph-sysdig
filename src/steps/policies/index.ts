import {
  createDirectRelationship,
  Entity,
  IntegrationStep,
  IntegrationStepExecutionContext,
  RelationshipClass,
} from '@jupiterone/integration-sdk-core';
import { createAPIClient } from '../../client';

import { IntegrationConfig } from '../../config';
import { ACCOUNT_ENTITY_KEY } from '../account';
import { Steps, Entities, Relationships } from '../constants';
import { createPolicyEntity } from './converter';

export async function fetchPolicies({
  instance,
  jobState,
}: IntegrationStepExecutionContext<IntegrationConfig>) {
  const apiClient = createAPIClient(instance.config);
  const accountEntity = (await jobState.getData(ACCOUNT_ENTITY_KEY)) as Entity;

  await apiClient.iteratePolicies(async (policy) => {
    const policyEntity = await jobState.addEntity(createPolicyEntity(policy));

    await jobState.addRelationship(
      createDirectRelationship({
        _class: RelationshipClass.HAS,
        from: accountEntity,
        to: policyEntity,
      }),
    );
  });
}

export const policiesSteps: IntegrationStep<IntegrationConfig>[] = [
  {
    id: Steps.POLICIES,
    name: 'Fetch Policies',
    entities: [Entities.POLICY],
    relationships: [Relationships.ACCOUNT_HAS_POLICY],
    dependsOn: [Steps.ACCOUNT],
    executionHandler: fetchPolicies,
  },
];
