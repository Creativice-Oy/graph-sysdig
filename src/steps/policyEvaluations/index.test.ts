import {
  createMockStepExecutionContext,
  Recording,
} from '@jupiterone/integration-sdk-testing';
import { IntegrationConfig } from '../../config';
import { fetchPolicyEvaluations } from '.';
import { integrationConfig } from '../../../test/config';
import { setupSysdigRecording } from '../../../test/recording';
import { fetchImageScans } from '../scans';
import { Relationships } from '../constants';
import { fetchPolicies } from '../policies';
import { fetchAccountDetails } from '../account';

describe('#fetchPolicyEvaluations', () => {
  let recording: Recording;

  beforeEach(() => {
    recording = setupSysdigRecording({
      directory: __dirname,
      name: 'fetchPolicyEvaluations',
    });
  });

  afterEach(async () => {
    await recording.stop();
  });

  test('should collect data', async () => {
    const context = createMockStepExecutionContext<IntegrationConfig>({
      instanceConfig: integrationConfig,
    });

    await fetchAccountDetails(context);
    await fetchPolicies(context);
    await fetchImageScans(context);
    await fetchPolicyEvaluations(context);

    expect({
      numCollectedEntities: context.jobState.collectedEntities.length,
      collectedEntities: context.jobState.collectedEntities,
      encounteredTypes: context.jobState.encounteredTypes,
    }).toMatchSnapshot();

    const policies = context.jobState.collectedEntities.filter(
      (e) => e._type === 'sysdig_policy',
    );
    expect(policies.length).toBeGreaterThan(0);
    expect(policies).toMatchGraphObjectSchema({
      _class: ['Policy'],
      schema: {
        additionalProperties: false,
        properties: {
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          _type: { const: 'sysdig_policy' },
          id: { type: 'string' },
          name: { type: 'string' },
          title: { type: 'string' },
          identifier: { type: 'string' },
          summary: { type: 'string' },
          content: { type: 'string' },
          creationTimestamp: { type: 'number' },
          updateTimestamp: { type: 'number' },
        },
      },
    });

    const imageScans = context.jobState.collectedEntities.filter((e) =>
      e._type.includes('sysdig_image_scan'),
    );
    expect(imageScans.length).toBeGreaterThan(0);

    if (imageScans.filter((scan) => !!scan.imageId).length > 0) {
      expect(
        imageScans.filter((scan) => !!scan.imageId),
      ).toMatchGraphObjectSchema({
        _class: ['Assessment'],
        schema: {
          additionalProperties: false,
          properties: {
            _rawData: {
              type: 'array',
              items: { type: 'object' },
            },
            _type: { const: 'sysdig_image_scan' },
            analysisStatus: { type: 'string' },
            analyzedAt: { type: 'number' },
            createdAt: { type: 'number' },
            fullTag: { type: 'string' },
            imageDigest: { type: 'string' },
            imageId: { type: 'string' },
            parentDigest: { type: 'string' },
            tagDetectedAt: { type: 'number' },
            registry: { type: 'string' },
            repository: { type: 'string' },
            tag: { type: 'string' },
            origin: { type: 'string' },
            policyStatus: { type: 'string' },
          },
        },
      });
    }

    if (imageScans.filter((scan) => !!scan.id).length > 0) {
      expect(
        imageScans.filter((scan) => !!scan.imageId),
      ).toMatchGraphObjectSchema({
        _class: ['Assessment'],
        schema: {
          additionalProperties: false,
          properties: {
            _rawData: {
              type: 'array',
              items: { type: 'object' },
            },
            _type: { const: 'sysdig_image_scan' },
            id: { type: 'string' },
            name: { type: 'string' },
            storedAt: { type: 'number' },
            type: { type: 'string' },
            'metadata.imageId': { type: 'string' },
            'metadata.pullString': { type: 'string' },
            'metadata.baseOS': { type: 'string' },
            'metadata.digest': { type: 'string' },
            'metadata.createdAt': { type: 'number' },
            'metadata.author': { type: 'string' },
            'metadata.size': { type: 'number' },
            'metadata.os': { type: 'string' },
            'metadata.architecture': { type: 'string' },
            'metadata.labels': { type: 'string' },
            'metadata.layersCount': { type: 'number' },
            vulnsBySev: { type: 'number' },
            packageCount: { type: 'number' },
            packageTypes: {
              type: 'array',
              items: { type: 'string' },
            },
            fixablePackages: {
              type: 'array',
              items: { type: 'any' },
            },
            runningFixablePackages: { type: 'any' },
            category: { type: 'string' },
            summary: { type: 'string' },
            internal: { type: 'boolean' },
          },
        },
      });
    }

    const policyEvaluations = context.jobState.collectedEntities.filter((e) =>
      e._type.includes('sysdig_policy_evaluation'),
    );
    expect(policyEvaluations.length).toBeGreaterThan(0);
    expect(
      imageScans.filter((scan) => !!scan.imageId),
    ).toMatchGraphObjectSchema({
      _class: ['Assessment'],
      schema: {
        additionalProperties: false,
        properties: {
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          _type: { const: 'sysdig_policy_evaluation' },
          id: { type: 'string' },
          name: { type: 'string' },
          category: { type: 'string' },
          summary: { type: 'string' },
          internal: { type: 'boolean' },
          completedOn: { type: 'number' },
          identifier: { type: 'string' },
          policyType: { type: 'string' },
          evaluationResult: { type: 'string' },
          'failuresCount.imageConfigCreationDate': { type: 'number' },
          'failuresCount.imageConfigDefaultUser': { type: 'number' },
          'failuresCount.imageConfigEnvVariable': { type: 'number' },
          'failuresCount.imageConfigLabel': { type: 'number' },
          'failuresCount.imageConfigSensitiveInformationAndSecrets': {
            type: 'number',
          },
          'failuresCount.vulnDenyList': { type: 'number' },
          'failuresCount.imageConfigInstructionNotRecommended': {
            type: 'number',
          },
          'failuresCount.vulnSeverityAndThreats': { type: 'number' },
          creationTimestamp: { type: 'number' },
          updateTimestamp: { type: 'number' },
        },
      },
    });

    expect(
      context.jobState.collectedRelationships.filter(
        (e) =>
          e._type === Relationships.POLICY_EVALUATION_REVIEWED_IMAGE_SCAN._type,
      ),
    ).toMatchDirectRelationshipSchema({
      schema: {
        properties: {
          _class: { const: 'REVIEWED' },
          _type: {
            const: 'sysdig_policy_evaluation_reviewed_image_scan',
          },
        },
      },
    });

    expect(
      context.jobState.collectedRelationships.filter(
        (e) =>
          e._type === Relationships.POLICY_EVALUATION_ENFORCES_POLICY._type,
      ),
    ).toMatchDirectRelationshipSchema({
      schema: {
        properties: {
          _class: { const: 'ENFORCES' },
          _type: {
            const: 'sysdig_policy_evaluation_enforces_policy',
          },
        },
      },
    });
  });
});
