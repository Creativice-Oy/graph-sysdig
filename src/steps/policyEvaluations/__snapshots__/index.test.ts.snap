// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#fetchImageScans should collect data 1`] = `
Object {
  "collectedEntities": Array [],
  "encounteredTypes": Array [],
  "numCollectedEntities": 0,
}
`;
